### ISSUE TITLE
Swipeable Bottom Sheet (new component)

### ISSUE THREAD

**Issue by slowaways** [quasarframework/quasar#7541]
**Describe the solution you'd like**
All the apps in the top 10 massively uses a Bottom Sheet Swipeable, where you can add any component to it, not just a menu of actions displayed as list or a grid.

I still don't know why Quasar doesn't have a Swipeable Bottom Sheet component like this:
https://github.com/atsutopia/vue-swipeable-bottom-sheet

**Describe alternatives you've considered**
---

**Additional context (examples)**
Instagram
![image](https://user-images.githubusercontent.com/9668277/89064529-891d7c00-d340-11ea-8bb4-4ccced059f9c.png)

Telegram:
![image](https://user-images.githubusercontent.com/9668277/89064549-90448a00-d340-11ea-9f98-faaf29f5a3a2.png)

OneDrive:
![doc-2020-07-31-12-57-30](https://user-images.githubusercontent.com/9668277/89064358-422f8680-d340-11ea-9f1c-6dbe1580a49b.gif)

Facebook:
![doc-2020-07-31-12-54-27](https://user-images.githubusercontent.com/9668277/89064508-802caa80-d340-11ea-97d8-b8c51468f017.gif)

**Comment by pdanpdan** [comment]
It's not so difficult to do it adjusted to your needs. It would take less than recording all these images :)

https://codepen.io/pdanpdan/pen/ExPqZbK

The problem is that it's hard to do it as a generic component because each use case is different.

**Comment by slowaways** [comment]
@pdanpdan,
Does your example work as a Bottom Sheet? The top bar that is used to swipe is never hidden, does it always appear there in the footer?
The fixed footer is only while the Bottom Sheet is active if you have a different layout: see the example of the Bottom Sheet that appears in the Comment on Facebook App (last GIF and first example).

Why not just to start with the current [Quasar Bottom Sheet plugin](https://quasar.dev/quasar-plugins/bottom-sheet) making it swipeable by setting a default minimal breakpoint height to auto dismiss it after release the touch?
Example: if the height reaches at least 50px when it is swipe down and the finger on the screen is released, hide the Bottom Sheet.

```
  this.$q.bottomSheet({
    message: 'Bottom Sheet message',
    swipeable: true,
    swipebar: true, // If swipebar is true, it is only possible to swipe using the top bar. If false, the entire Bottom Sheet is swipeable.
    breakpoint: 50 // default value
  })
```

I think the easiest solution would be to add an prop to make a Dialog with position="bottom" swipeable by setting some directive for this in QHeader if it had a QLayout for example.  So it is possible to compose this fake "Bottom Sheet" with any component and still play with the most varied layouts: fixed footer, etc.

**Comment by pdanpdan** [comment]
Why not just tweak the example not to show the bar? It's all there, just that exactly how I said everyone wants something else.

**Comment by slowaways** [comment]
I checked all the code.

I believe that we are discussing a solution to be implemented within the Quasar core and not a specific solution for me.

I just thought that you had confused the fixed footer, so I tried to clarify that this is the example of the last GIF, an action to comment there on FB.

**Comment by IlCallo** [comment]
@slowaways what you are referring to is a mix of [backdrop](https://material.io/components/backdrop) and [bottom sheet](https://material.io/components/sheets-bottom) from material design spec. Both are relatively recent (especially the former) and the version with the drag-bar isn't even officially mentioned AFAIK, even if it has been used into Google Maps for years .

I also checked other material-based UI libraries and could not find a similar component, feel free to provide resources on the topic.

I like the component idea, but it should probably be added as a standalone Quasar App Extension until the spec is settled and stable.
Would you like to give a try? :)

**Comment by slowaways** [comment]
@IlCallo Why does Quasar have to strictly follow the specification?
This is an overly important and massively used component. I don't know why there is no specification for it.

I'll try to work on it through the Dialog code base. Just try to add the "Swipeable" part.

**Comment by pdanpdan** [comment]
@slowaways is it so difficult for you to do basic adjustments to an already complete component to adjust it as you want it?

**Comment by slowaways** [comment]
@pdanpdan That's what I'm going to do.
I'm just going to create the "Swipeable" feature within the dialog component.
But I intend to send PR to the Quasar repository if it looks good.
I intend to use this component in several of my projects.

**Comment by IlCallo** [comment]
> Why does Quasar have to strictly follow the specification?

It doesn't have to, but following a spec makes it easier to maintain the codebase :)

Unstable or missing specs means that at some point we would probably need to change the component implementation or aesthetic, and this would generate a breaking change.
This in turn would force us to:
- just break every existing apps using the previous version (very bad)
- release a new major version (still bad, we can't do it for every breaking change and we must accumulate them in batches)
- support both the old and the new as opt-in (bad on our side, we must maintain the same code twice)

This is just one of the reasons following specs is a good (and future proof) idea, there are many others.

But if you write an app-extension providing this component and it gets major traction, we could consider to integrate it into the core UI.
This is the compromise to allow everyone to get the component they want and still keep the framework stable and easy to maintain

**Comment by IlCallo** [comment]
@slowaways have you started working on this?
I may create it as an UI AE in next few weeks, if you started and have some advices, please share

**Comment by slowaways** [comment]
> @slowaways have you started working on this?
> I may create it as an UI AE in next few weeks, if you started and have some advices, please share

I have not started yet.

To make the functionality look like the examples above, I thought of a new property in the Touch Pan Directive:
`handlePan ({ evt, ...info }) {}`
`{{ info.speed }}` -> `10` (pixel per second)
A property called "speed", which measures the number of pixels per second. With that, it was possible to detect if the user wants to dismiss the component, quickly swipe it down. For this to work, it was only necessary to create a logic that defined a speed limit to, if it exceeds that limit, automatically dispense the component.

**Comment by pdanpdan** [comment]
cross reference:
https://github.com/quasarframework/quasar/issues/7541
https://github.com/quasarframework/quasar/issues/5202

**Comment by bake09** [comment]
@pdanpdan @slowaways 
Thank you for the codepen Example: 
I forked your Pen and try to add a "click-outside to close" functionallity...
- I used a additional div at the bottom 
- add z-index to the card and div.
- try to give a dynamic background transistion due rgba

@pdanpdan Can you give a look please?
is this good practice?

https://codepen.io/bake09/pen/jOGLyya

Thanks very much

**Comment by pdanpdan** [comment]
LGTM, just a few tweaks: https://codepen.io/pdanpdan/pen/yLzogKq?editors=1010
- hide backdrop when in handler mode
- order backdrop first and drawer second so you don't need zIndex

**Comment by douglasg14b** [comment]
Really just a bottom sheet that works like the side nav does, or just a base component that has the same swipeable behavior as the side nav drawer, but in any direction would be nice...

**Comment by FadhiliNjagi** [comment]
The component being described here is a draggable bottom sheet (with a drag handle). It's now part of the Material Design 3 spec [https://m3.material.io/components/bottom-sheets/overview](https://m3.material.io/components/bottom-sheets/overview) .

I keep seeing them on Twitter (the website and PWA) on mobile and I'm like how can I add this to my next UI!

**Comment by DaleMckeown** [comment]
Yes, now that this is part of Material Design 3 spec, hopefully this will be implemented in Quasar?

**Comment by AntonERG** [comment]
Just in case anybody will find it helpful, I occasionally found this implementation (without any extra plugins):
https://codepen.io/pdanpdan/pen/ExPqZbK

**Comment by ptc-tonyo** [comment]
This addition to Quasar is IMHO now *long* overdue ... or at least as a plugin.
I'd use it in almost every project as its very commonly used and expected now. I'm sure I'm not alone on this!

Options: 
- 1 or 2 height snap points
- Bottom snap point can be 0 => Completely hide and thus be triggered by swipe up only
- Show standard handle at the top (thickness / width)
- Default slot that works like QDialog with optional tap outside to close or swipe down to close.

This Vue example feels right but uses HammerJS for touch events.
- https://github.com/vaban-ru/vue-bottom-sheet
